src/core/Form/TextInput/TextInput.tsx

```diff
@@ -78,15 +121,62 @@ export class TextInput extends Component<TextInputProps> {
     };
 
     return (
-      <StyledTextInput
-        {...passProps}
-        className={classnames(className, {
+      <HtmlDiv
+        {...inputContainerProps}
+        className={classnames(baseClassName, className, {
+          [textInputClassNames.disabled]: !!passProps.disabled,
           [textInputClassNames.icon]: resolvedIcon !== undefined,
+          [textInputClassNames.error]: status === 'error',
+          [textInputClassNames.success]: status === 'success',
+          [textInputClassNames.fullWidth]: fullWidth,
         })}
       >
-        {children}
-        {resolvedIcon && <Icon {...newIconProps} />}
-      </StyledTextInput>
+        <LabelText
+          htmlFor={this.id}
+          labelMode={labelMode}
+          as="label"
+          optionalText={optionalText}
+        >
+          {labelText}
+        </LabelText>
+        <HintText id={this.hintTextId}>{hintText}</HintText>
+        <HtmlDiv className={textInputClassNames.functionalityContainer}>
```

Comment:
Functionality container is probably unnecessary complexity for text input. This was introduced for search input because of visual glitches in shadows when focus styles were applied to multiple components (input, clear button, search button). Should be possible to combine functionalityContainer and inputElementContainer for this component.

Reply:
Good find.
This `functionalityContainer` is now removed.